pilot/pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leaderelection
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	coreV1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	electionId = "istio-leader"
+)
+
+type LeaderElection struct {
+	namespace string
+	name      string
+	runFns    []func(stop <-chan struct{})
+	client    kubernetes.Interface
+}
+
+// Run will start leader election, calling all runFns when we become the leader.
+func (l *LeaderElection) Run(stop <-chan struct{}, onComplete func()) error {
```

Comment:
All of pilot is doing the stop channel passing.  So at some point, we will need to convert stop channel into context? So I think its reasonable to do here rather than outside this function?

I feel like there should be a simpler way though..